rag/advanced_rag/agentic_rag.py

```
@@ -653,36 +661,41 @@ async def rag(self, question: str) -> str:
+        # object (bound through the per-task ContextVar), so parallel calls
+        # report independent usage instead of one shared cumulative sink.
+        with using_stats(LLMUsageStats()) as call_stats:
+            # P0: reuse a near-identical question's cached answer instead of re-running
+            # the whole agentic graph. Significant-keyword overlap (>= min_overlap AND
+            # >=2 shared words, and matching numbers) collapses the re-ask pattern
+            # while leaving genuinely different questions untouched. Attachments bypass
+            # the cache (their content is appended to the question message below).
+            if question and not self.text_attachments_content:
+                qk = _question_keywords(question)
+                if self._rag_cache:
+                    for cached_q, (cached_answer, cached_gram) in list(self._rag_cache.items()):
```

Ruff (0.16.1)

[warning] Loop control variable `cached_q` not used within loop body

Rename unused `cached_q` to `_cached_q`

(B007)